Keyboard handler rewrite

Spengreb opened the pull request:
I rewrote how LUPPP handles the keyboard shortcuts. This pull request adds a new context menu button when right clicking on a clip. It shows the current hotkey assigned to this clip and allows the user to change it with a single button press. If another hotkey is already assigned somewhere else it will remove that assignment applying it to only the current clip in context. 

Heres what it looks like:

![image](https://github.com/openAVproductions/openAV-Luppp/assets/28860841/790e4cf7-274e-4eef-a640-3000ceea4601)

This addresses quite a few problems from the previous implementation. One thing was that the hotkeys are not communicated in any way to the user and when first using LUPPP it was a little confusing, especially when not using a US/UK qwerty keyboard. 

As it stands in this PR the new hotkeys are only set for that instance of the application. I want this to persist from one luppp launch to the other and I would like to make the other special keys configurable too in the `setup` menu. To prevent this branch from ballooning too much I think I'm at a point where its good enough to merge.

harryhaaren commented on src/diskreader.cxx:
```diff
@@ -93,6 +93,23 @@ int DiskReader::loadPreferences()
 			LUPPP_NOTE("No default controllers active.");
 		}
 
+		cJSON* hotkeyAssignment = cJSON_GetObjectItem(preferencesJson, "hotkeyAssignment");
+        if (hotkeyAssignment) {
+            keyToGrid.clear();
+            cJSON* key = hotkeyAssignment->child;
+            while (key) {
+                int keycode = std::stoi(key->string);
+                cJSON* gridArray = key;
+                if (cJSON_GetArraySize(gridArray) == 2) {
+                    int x = cJSON_GetArrayItem(gridArray, 0)->valueint;
+                    int y = cJSON_GetArrayItem(gridArray, 1)->valueint;
+                    keyToGrid[keycode] = {x, y};
+                }
+                key = key->next;
+            }
+        } else {
+            LUPPP_WARN("No hotkeyAssignment found in preferences");
+        }
```
The indentation here looks off? Perhaps there's a tabs/spaces mixup?

Spengreb replied:
Hopefully this is sorted now. This wasnt clear in vs code but is in clion. I fixed some more areas where i messed this up.

harryhaaren commented on src/diskreader.cxx:
```diff
@@ -93,6 +93,23 @@ int DiskReader::loadPreferences()
+            keyToGrid.clear();
+            cJSON* key = hotkeyAssignment->child;
+            while (key) {
+                int keycode = std::stoi(key->string);
```
is there any way this can throw an exception due to invalid/non-int-convertable input? `stoi` could[1], so this needs a try {} catch to avoid taking the whole program down on int-convert-error?

Perhaps the `C` lang `atoi` (with no exceptions, and no error handling.. returning `0` instead) is better as it won't throw exceptions?

[1] https://en.cppreference.com/w/cpp/string/basic_string/stol

harryhaaren reviewed (changes requested):
try {} catch {} or `atoi` for avoiding C++ exception throwing. Elegant solution (much more so than my "i'm learning C++ so duplicating code everywhere" approach :)

Spengreb commented:
The above comments should all be resolved. I also did a bit of a refactor of the hotkey dialog box as it was a bit messy before. I fixed some bugs i noticed along the way also. Such as not allowing SHIFT and ESC to be set as a hotkey. Since SHIFT is the inverse of an action, it caused some very confusing behavior if allowed to be set as a hotkey as well. 

I also changed the format of hotkey dialog on the front end to try to communicate some of this to the user here is what it looks like now:

![image](https://github.com/user-attachments/assets/046cd54d-e76d-4061-8962-80a75df212fb)
